### PR TITLE
chore(button): drop classic support FE-2629

### PR DESCRIPTION
**Pull request by jamime**
### Proposed behaviour
Remove the Classic theme from `Button` to make adding new variants easier.
The Classic button is also used by `Split Button` and `Multi-Action Button`, these components have also had Classic theme removed.

### Current behaviour
`Button`, `Split Button` and `Multi-Action Button` has a classic Theme.

image
image
image
image
image
image
image
image
image
image
image
image
image
image

image
image

image
image

### Checklist

- [x] Release notes (Conventional Commits) 
- [x] Unit tests
- [x] Cypress automation tests
- [x] Storybook added or updated
<del>- [ ] Theme support
<del>- [ ] Typescript `d.ts` file added or updated
### Additional context
I plan on revisiting Button again to make further improvements. This PR is the minimum changes required to remove Classic from Button. Doing so will make it significantly easier to add more variations. 

In a future PR I hope to
- rename the `as` prop to `buttonType`
- simplify button styling
- remove react-router
- prevent `<a type="button" />` when passing a `href`

### Testing instructions

There will be other Classic components that use Button including `Form` and `Dialog` what will now display a Base/Mint/Aegean button instead of Classic. This is to be expected. We no longer maintain the Classic theme for any components and as such our consumers should be migrating to a "modern" theme.

Use `npm start` to start Storybook.

The following stories should **look** the same.
_Please note, all of the knobs are now available, previously some were conditional based on if you had a different knob value. The functionality is the same._

| Local | Production |
|-|-|
| [Button Default][button] | [Button Default][button-prod] |
| [Multi-Action Button Default][multi-action-button] | [Multi-Action Button Default][multi-action-button-prod] |
| [Split Button Default][split-button] | [Split Button Default][split-button-prod] |

The following stories have been removed

| Local | Production |
|-|-|
| [Button Classic][button-classic] | [Button Classic][button-classic-prod] |
| [Button as a Sibling Classic][button-sibling-classic] | [Button as a Sibling Classic][button-sibling-classic-prod] |
| [Split Button Classic][split-button-classic] | [Split Button Classic][split-button-classic-prod] |
| [Multi-Action Button Classic][multi-action-button-classic] | [Multi-Action Button Classic][multi-action-button-classic-prod] |

[button]: http://localhost:9001/?path=/story/button--default
[button-prod]: https://carbon.sage.com/?path=/story/button--default
[button-classic]: http://localhost:9001/?path=/story/button--classic
[button-classic-prod]: https://carbon.sage.com/?path=/story/button--classic
[button-sibling-classic]: http://localhost:9001/?path=/story/button--as-a-sibling-classic
[button-sibling-classic-prod]: https://carbon.sage.com/?path=/story/button--as-a-sibling-classic

[multi-action-button]: http://localhost:9001/?path=/story/multi-action-button--default
[multi-action-button-prod]: https://carbon.sage.com/?path=/story/multi-action-button--default
[multi-action-button-classic]: http://localhost:9001/?path=/story/multi-action-button--classic
[multi-action-button-classic-prod]: https://carbon.sage.com/?path=/story/multi-action-button--classic

[split-button]: http://localhost:9001/?path=/story/split-button--default
[split-button-prod]: https://carbon.sage.com/?path=/story/split-button--default
[split-button-classic]: http://localhost:9001/?path=/story/split-button--classic
[split-button-classic-prod]: https://carbon.sage.com/?path=/story/split-button--classic

### CODE REVIEW

**Comment by jamime** [comment]
Removing the `classic` theme for button includes deprecating the `as` prop. This is used internally. I'm going to create a codemod to assist others with the migration.

File: src/components/button/button-types.style.js [hunk 1]
```diff
@@ -9,92 +9,93 @@ export function makeColors(color) {
...
-    &:hover {
-      background: ${colors.destructive.hover};
-    }
+export default ({
+  theme: { colors, disabled }, disabled: isDisabled, destructive, buttonType
+}) => {
+  const styles = {
+    primary: `
+      background: ${colors.primary};
+      border-color: transparent;
```
Comment (grabkowski) [inline src/components/button/button-types.style.js]: I think that the `border-color` is already transparent in the main file

File: src/components/button/button-types.style.js [hunk 2]
```diff
@@ -9,92 +9,93 @@ export function makeColors(color) {
...
-  secondary: `
+        &:hover {
+          background: ${disabled.button};
+          border-color: ${disabled.button};
+          color: ${disabled.text};
+        }
+      `)}      
+    `,
+    secondary: `
+        background: transparent;
```
Comment (grabkowski) [inline src/components/button/button-types.style.js]: isn't the background transparent by default?

File: src/components/button/button-types.style.js [hunk 3]
```diff
@@ -9,92 +9,93 @@ export function makeColors(color) {
...
+      border-color: transparent;
+      color: ${colors.white};
+      &:hover {
+        background: ${colors.secondary};
+      }
+  
+      ${IF(destructive, `
+      background: ${colors.error};
+      border-color: transparent;
+      color: ${colors.white};
```
Comment (grabkowski) [inline src/components/button/button-types.style.js]: that does nothing, as the border color is already transparent and text is already white

File: src/components/button/button-types.style.js [hunk 4]
```diff
@@ -9,92 +9,93 @@ export function makeColors(color) {
...
+      &:hover {
+        background: ${colors.destructive.hover};
+      }
+      `)}
```
Comment (grabkowski) [inline src/components/button/button-types.style.js]: could you indent these styles to increase readability?

File: src/components/button/button-types.style.js [hunk 5]
```diff
@@ -9,92 +9,93 @@ export function makeColors(color) {
...
-      &:hover {
+  
+      ${IF(isDisabled, `
         background: ${disabled.button};
-        ${makeColors(disabled.text)}
-      }
-    ` : ''}
-  `
-});
+        color: ${disabled.text};
```
Comment (grabkowski) [inline src/components/button/button-types.style.js]: this color is common to all disabled buttons, maybe we could have it directly in the main file?

File: src/components/button/button-types.style.js [hunk 6]
```diff
@@ -9,92 +9,93 @@ export function makeColors(color) {
...
+        &:hover {
+          background: ${disabled.button};
+          ${makeColorAndIconColor(disabled.text)}
```
Comment (grabkowski) [inline src/components/button/button-types.style.js]: this is also common to all disabled buttons except primary which does not have the`${StyledIcon}` color. I think that could  be also set directly in the main file if that icon color does not matter.

File: src/components/button/button.spec.js [hunk 7]
```diff
@@ -177,67 +175,11 @@ describe('Button', () => {
...
-  });
-
-  describe('when not in classic theme', () => {
-    it('matches the applies the expected style to the icon', () => {
-      const wrapper = TestRenderer.create(<StyledButton iconType='plus' />);
-      assertStyleMatch({
-        height: '16px'
-      }, wrapper.toJSON(), { modifier: `${StyledIcon}` });
-    });
+  it('matches the applies the expected style to the icon', () => {
```
Comment (edleeks87) [inline src/components/button/button.spec.js]:
```suggestion
  it('applies the expected style to the icon', () => {
```

Comment (edleeks87) [reply src/components/button/button.spec.js]: or matches if that fits better

Comment (jamime) [reply src/components/button/button.spec.js]: This is just a refactor of an existing test. I'll look to tidy the test suite in a future PR.

File: src/components/button/button.stories.js [hunk 8]
```diff
@@ -2,9 +2,9 @@ import React from 'react';
 import { storiesOf } from '@storybook/react';
 import { text, select, boolean } from '@storybook/addon-knobs';
 import { action } from '@storybook/addon-actions';
-import { dlsThemeSelector, classicThemeSelector } from '../../../.storybook/theme-selectors';
+import { dlsThemeSelector } from '../../../.storybook/theme-selectors';
```
Comment (grabkowski) [inline src/components/button/button.stories.js]: do we need that `dlsThemeSelector`? Looks like it's an empty object.

Comment (jamime) [reply src/components/button/button.stories.js]: We do not. Resolved in a8a2ac277d632d9784b013f3cec0876fd99af73f

**Comment by DlgSHi** [comment]
Looks ok for me, just need to be removed also `buttonClassic.feature`, `buttonAsSiblingClassic.feature`, `splitButtonClassic.feature`.
We have 2 options:
- delete those files in this PR
- wait until this [PR](https://github.com/Sage/carbon/pull/2751) about deleting all classic tests will be merged

**Comment by jamime** [comment]
We'll get your PR merged first :)

**Pull request merged**